airbyte-cdk/python/airbyte_cdk/entrypoint.py

```diff
@@ -160,8 +161,36 @@ def read(
         if self.source.check_config_against_spec:
             self.validate_connection(source_spec, config)
 
-        yield from self.source.read(self.logger, config, catalog, state)
-        yield from self._emit_queued_messages(self.source)
+        stream_message_counter: DefaultDict[HashableStreamDescriptor, int] = defaultdict(int)
+        for message in self.source.read(self.logger, config, catalog, state):
+            yield self.handle_record_counts(message, stream_message_counter)
+        for message in self._emit_queued_messages(self.source):
+            yield self.handle_record_counts(message, stream_message_counter)
+
+    @staticmethod
+    def handle_record_counts(message: AirbyteMessage, stream_message_count: DefaultDict[HashableStreamDescriptor, int]) -> AirbyteMessage:
+        if message.type == Type.RECORD:
+            stream_message_count[message_utils.get_stream_descriptor(message)] += 1
+
+        elif message.type == Type.STATE:
+            if not message.state.stream:
+                raise ValueError("State message was not in per-stream state format, which is required for record counts.")
+
+            stream_descriptor = message_utils.get_stream_descriptor(message)
+
+            # Set record count from the counter onto the state message
+            message.state.sourceStats = message.state.sourceStats or AirbyteStateStats()
```

Comment:
Is that the same as writing the following?

```python
if message.state.sourceStats is None:
  message.state.sourceStats = AirbyteStateStats()
```

Do we universally prefer the ` | ` format?

Reply:
At least for python i prefer `or` because its less verbose and pretty readable